### PR TITLE
Post-hackathon changes

### PR DESCRIPTION
**Pull request by merklejerk**
Now that the comp has wrapped up, here's a couple things I've been wanting to add/change. 

- The biggest feature change is the addition of a payment verification function `verifyBountyCanPay()`, which is a sandbox around payment alone. Because the system is non-custodial, the only way to check if a bounty can pay is to actually simulate it in a sandbox. 
  - Lots of refactoring has been done to accommodate this. 
- Operators can also update all fields of their bounty to make it easier/cheaper to adjust bounties as a project evolves. 
- Loads more tests to get up to 97% line coverage.

### CODE REVIEW

**Review by JordanCason — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1,71 +1,71 @@
 # HoneyPause
 
-HoneyPause is a permissionless on-chain exploit bounty tied to a circuit breaker. HoneyPause lets whitehats safely and atomically prove a smart contract exploit <i>on-chain</i>, pause the affected protocol, then collect a bounty. Protocols can opt into the system by registering a bounty on the smart contract. The entire system is permissionless, non-custodial, and free!
+HoneyPause is an onchain exploit bounty linked to a circuit breaker. HoneyPause lets whitehats safely and atomically prove a smart contract exploit on chain, **pause** the affected protocol, then collect a bounty-- all in a single transaction. Projects opt into the system by registering a bounty on the smart contract. The entire system is permissionless, non-custodial, and free!
```
Comment (justinschuldt) [inline README.md:3]:
```suggestion
HoneyPause is an onchain exploit bounty linked to a circuit breaker. HoneyPause lets whitehats safely and atomically prove a smart contract exploit on chain, `pause` the affected protocol, then collect the bounty-- all in a single transaction. Projects opt into the system by registering a bounty on the smart contract. The entire system is permissionless, non-custodial, and free!
```

File: README.md [hunk 2]
```diff
@@ -1,71 +1,71 @@
...
 
-For Ethereum applications that can be exploited in a single transaction, this adds another form of proactive defense that can supplement traditional (off-chain) bug bounties and monitoring.
+For Ethereum protocols that can be exploited in a single transaction (which are often the case), this offers a novel form of proactive defense that can complement traditional (off-chain) bug bounties, threat monitoring, and in-protocol circuit breakers.
```
Comment (justinschuldt) [inline README.md:5]:
```suggestion
For Ethereum protocols that can be exploited in a single transaction (which are often the case), this offers a novel form of defense that can complement traditional (off-chain) bug bounties, threat monitoring, and in-protocol circuit breakers.
```

File: README.md [hunk 3]
```diff
@@ -1,71 +1,71 @@
...
+## How it Works
 
 ### Protocol Registration
-First a protocol must register itself to the HoneyPause contract via the `add()` function, providing:
+Projects register themselves to the HoneyPause contract via the `add()` function, providing:
 
-1. A bounty token and amount.
-2. The address of a custom **Verifier** contract which asserts on-chain state invariants that would be violated in the case of an exploit. Examples could be an AMM's reserve violating the constant product formula or a lending protocol incurring bad debt.
+1. A bounty token and amount, but no deposit*.
+2. The address of a custom **Verifier** contract which asserts onchain state invariants that would be violated in the case of an exploit. Examples could be an AMM's reserve violating the constant product formula or a lending protocol incurring bad debt.
```
Comment (justinschuldt) [inline README.md:13]:
```suggestion
2. The address of a custom `Verifier` contract which asserts onchain state invariants that would be violated in the case of an exploit. Examples could be an AMM's reserve violating the constant product formula or a lending protocol incurring bad debt.
```

File: README.md [hunk 4]
```diff
@@ -1,71 +1,71 @@
...
 3. The address of a custom **Pauser** contract which is authorized to pause/freeze the protocol when called by the HoneyPause contract.
```
Comment (justinschuldt) [inline README.md:14]:
```suggestion
3. The address of a custom [`Pauser`](./src/HoneyPause.sol#L28) contract which is authorized to pause/freeze the protocol when called by the HoneyPause contract.
```

File: README.md [hunk 5]
```diff
@@ -1,71 +1,71 @@
...
 4. The address of a custom **Payer** contract which must pay the bounty to the whitehat when called by the HoneyPause contract.
```
Comment (justinschuldt) [inline README.md:15]:
```suggestion
4. The address of a custom [`Payer`](./src/HoneyPause.sol#L42) contract which must pay the bounty to the whitehat when called by the HoneyPause contract. 
```

File: README.md [hunk 6]
```diff
@@ -1,71 +1,71 @@
...
+5. An operator account for the bounty, who will be able to modify the bounty.
 
-Note that the HoneyPause contract never custodies bounties. It is up to the protocol's **Payer** contract to surface funds to cover the bounty when called. This means a protocol may employ an indirect way of paying the bounty when it is demanded, such is liquidating assets, activating a safety module, etc.
+> \* Note that the HoneyPause contract never custodies bounties. It is up to the project's **Payer** contract to surface funds to cover the bounty when called.
 
 ### Claiming a Bounty
-A whitehat that has discovered an exploit on a registered protocol will post a `claim()` transaction **TO A PRIVATE MEMPOOL**, providing an **Exploiter** contract that will perform the exploit when called by the HoneyPause contract. The HoneyPause contract will:
+A whitehat that has discovered an exploit on a registered project will submit a successful `claim()` transaction **TO A PRIVATE MEMPOOL**, providing an **Exploiter** contract that will perform the exploit when called by HoneyPause. In the same transaction, the HoneyPause contract will:
```
Comment (justinschuldt) [inline README.md:21]:
```suggestion
A whitehat that has discovered an exploit on a registered project will submit a successful `claim()` transaction **TO A PRIVATE MEMPOOL**, providing an [`Exploiter`](./src/HoneyPause.sol#L35) contract that will perform the exploit when called by HoneyPause. In the same transaction, the HoneyPause contract will:
```

File: src/HoneyPause.sol [hunk 7]
```diff
@@ -99,13 +98,13 @@ contract HoneyPause {
...
+        return getBounty[bountyId].operator == address(0);
     }
 
     /// @notice Add a new bounty for a protocol.
     /// @param name Name of the protocol to emit as part of the listing.
     /// @param payoutToken The token the bounty will be paid out in. 0 for ETH.
-    /// @param payoutToken The bounty amount (in wei).
+    /// @param payoutAmount The bounty amount (in wei).
     /// @param verifier The protocol's verifier contract, which asserts that critical invariants have been broken.
     /// @param pauser A privileged contract that can pause the protocol when called.
```
Comment (justinschuldt) [inline src/HoneyPause.sol:109]:
```suggestion
    /// @param pauser A privileged contract that can pause the protocol when called.
    /// @param payer A contract that can pay the bounty when called.
```

File: src/HoneyPause.sol [hunk 8]
```diff
@@ -140,18 +133,35 @@ contract HoneyPause {
...
     /// @dev    Must be called by the current operator of a bounty.
-    /// @param bountyId The bounty.
-    /// @param newOperator The new operator address.
-    function replaceOperator(uint256 bountyId, address newOperator)
-        external onlyBountyOperator(bountyId)
+    /// @param bountyId ID of an existing bounty.
+    /// @param payoutToken The token the bounty will be paid out in. 0 for ETH.
+    /// @param payoutAmount The bounty amount (in wei).
+    /// @param verifier The protocol's verifier contract, which asserts that critical invariants have been broken.
+    /// @param pauser A privileged contract that can pause the protocol when called.
```
Comment (justinschuldt) [inline src/HoneyPause.sol:142]:
```suggestion
    /// @param pauser A privileged contract that can pause the protocol when called.
    /// @param payer A contract that can pay the bounty when called.
```

**Review by justinschuldt — COMMENTED** [review]
LG. Most of my suggestions are nits, feel free to dismiss

File: README.md (hunk 2 above)
Comment (merklejerk) [reply README.md:5]: y u hate proactive

**Pull request merged**